From 34784cc3809b4443a212935ede44169d3829749a Mon Sep 17 00:00:00 2001
From: Cristy <urban-warrior@imagemagick.org>
Date: Sat, 8 Jul 2017 18:32:55 -0400
Subject: [PATCH] memory leak in ReadOneJNGImage #550

A crafted file could trigger a memory leak

bug-debian: https://bugs.debian.org/870108
bug: https://github.com/ImageMagick/ImageMagick/issues/550
origin: https://github.com/ImageMagick/ImageMagick/commit/3320955045e5a2a22c13a04fa9422bb809e75eda,
 https://github.com/ImageMagick/ImageMagick/commit/982d89a952e7d6840ec7851c364f489a50d805b7

(cherry picked from commit 3320955045e5a2a22c13a04fa9422bb809e75eda and 982d89a952e7d6840ec7851c364f489a50d805b7)
---
 coders/png.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/coders/png.c b/coders/png.c
index bd740ecc1..2a27659f1 100644
--- a/coders/png.c
+++ b/coders/png.c
@@ -4166,7 +4166,13 @@ static Image *ReadOneJNGImage(MngInfo *mng_info,
         type[0],type[1],type[2],type[3],(double) length);
 
     if (length > PNG_UINT_31_MAX || count == 0)
-      ThrowReaderException(CorruptImageError,"CorruptImage");
+      {
+        if (color_image != (Image *) NULL)
+          color_image=DestroyImage(color_image);
+        if (color_image_info != (ImageInfo *) NULL)
+          color_image_info=DestroyImageInfo(color_image_info);
+        ThrowReaderException(CorruptImageError,"CorruptImage");
+      }
 
     p=NULL;
     chunk=(unsigned char *) NULL;
